### PR TITLE
add tier datasource, datasource filter, id string validator, tests

### PR DESCRIPTION
**Pull request by davidbloss**
## Changelog

Changes:
- Add `tier` datasource with tests.
- Add `FilterModel`. This data model holds data sourced from `filter` blocks in schemas. This `filter` block is used in multiple datasources - [tier example here](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/data-sources/tier)
- Add `IdStringValidator`. This is unused here, but will be useful for managed resources where the id is provided and not computed.
- `opselevel-go` version is updated

- [X] List your changes here
- [ ] Make a `changie` entry

## Tophatting

Using this file:
```terraform
# main.tf
data "opslevel_tier" "tier1_by_alias" {
  filter {
    field = "alias"
    value = "tier_1"
  }
}
data "opslevel_tier" "tier2_by_id" {
  filter {
    field = "id"
    value = "Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzg"
  }
}
data "opslevel_tier" "tier3_by_index" {
  filter {
    field = "index"
    value = 3
  }
}
data "opslevel_tier" "tier4_by_name" {
  filter {
    field = "name"
    value = "Tier 4"
  }
}

output "found_tier1_by_alias" {
  value = data.opslevel_tier.tier1_by_alias
}

output "found_tier2_by_id" {
  value = data.opslevel_tier.tier2_by_id
}

output "found_tier3_by_index" {
  value = data.opslevel_tier.tier3_by_index
}

output "found_tier4_by_name" {
  value = data.opslevel_tier.tier4_by_name
}
```

Run `terraform plan`
```terraform
data.opslevel_tier.tier3_by_index: Reading...
data.opslevel_tier.tier1_by_alias: Reading...
data.opslevel_tier.tier4_by_name: Reading...
data.opslevel_tier.tier2_by_id: Reading...
data.opslevel_tier.tier4_by_name: Read complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvVGllci8yNDA]
data.opslevel_tier.tier1_by_alias: Read complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzc]
data.opslevel_tier.tier3_by_index: Read complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzk]
data.opslevel_tier.tier2_by_id: Read complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzg]

Changes to Outputs:
  + found_tier1_by_alias = {
      + alias  = "tier_1"
      + filter = {
          + field = "alias"
          + value = "tier_1"
        }
      + id     = "Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzc"
      + index  = 1
      + name   = "Tier 1"
    }
  + found_tier2_by_id    = {
      + alias  = "tier_2"
      + filter = {
          + field = "id"
          + value = "Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzg"
        }
      + id     = "Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzg"
      + index  = 2
      + name   = "Tier 2"
    }
  + found_tier3_by_index = {
      + alias  = "tier_3"
      + filter = {
          + field = "index"
          + value = "3"
        }
      + id     = "Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzk"
      + index  = 3
      + name   = "Tier 3"
    }
  + found_tier4_by_name  = {
      + alias  = "tier_4"
      + filter = {
          + field = "name"
          + value = "Tier 4"
        }
      + id     = "Z2lkOi8vb3BzbGV2ZWwvVGllci8yNDA"
      + index  = 4
      + name   = "Tier 4"
    }

You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.
```

### CODE REVIEW

File: opslevel/validators.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+// ValidateString performs the validation.
+func (v idStringValidator) ValidateString(ctx context.Context, request validator.StringRequest, response *validator.StringResponse) {
+	if request.ConfigValue.IsNull() || request.ConfigValue.IsUnknown() {
+		return
+	}
+
+	value := request.ConfigValue.ValueString()
+
+	if opslevel.IsID(value) {
```
Comment (taimoorgit) [inline opslevel/validators.go:32]: Is this condition wrong?

Comment (davidbloss) [reply opslevel/validators.go:32]: Yes, this should be inverted. Good catch 👍

File: opslevel/datasource_opslevel_tier.go [hunk 2]
```diff
@@ -1,94 +1,137 @@
...
-// 	d.Set("name", item.Name)
-
-// 	return nil
-// }
+func filterTiers(tiers []opslevel.Tier, filter FilterModel) (*opslevel.Tier, error) {
+	if filter.Value.Equal(types.StringValue("")) {
+		return nil, fmt.Errorf("Please provide a non-empty value for filter's value")
+	}
+	for _, tier := range tiers {
+		switch filter.Field.ValueString() {
```
Comment (taimoorgit) [inline opslevel/datasource_opslevel_tier.go:115]: Should this do anything in the default case?

**Review by taimoorgit — APPROVED** [review]

File: opslevel/datasource_opslevel_tier.go (hunk 2 above)
Comment (davidbloss) [reply opslevel/datasource_opslevel_tier.go:115]: No need for a default case here.
Assuming the filter looks like:
```terraform
filter {
  field = "alias"
  value = "tier_1"
}
```
We want to check all tiers for the alias `tier_1`.
If this is the last item in `tiers`, we don't need to log anything along the way. If nothing is found after all tiers are looked through then the error below catches that case.

**Pull request merged**